### PR TITLE
Feat/financial-date-functions

### PR DESCRIPTION
**Pull request by RHDZMOTA**
This pull request provides the following functionality:
- Time object for different time-types (integer and date-time).
- Financial functions (xirr and xnpv) for cash flows subject to a particular date.
- Implementation of xirr and xnpv in the main cashflow object.

### CODE REVIEW

File: .idea/other.xml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="PySciProjectComponent">
+    <option name="PY_SCI_VIEW" value="true" />
+    <option name="PY_SCI_VIEW_SUGGESTED" value="true" />
+  </component>
+</project>
```
Comment (joseluisvaz) [inline .idea/other.xml]: Add .idea to .gitignore

File: README.md [hunk 2]
```diff
@@ -18,35 +18,27 @@ Check out the file **examples.ipynb** in order to see usage examples.
...
-sudo apt install python3-nose
-nosetests3
-```
+* Install the linux package: `sudo apt install python3-nose`
+* Install via pip: `pip install nose`
+In the base directory run the following:
 
-#### Python package
-Use the virtualenv with the required packages.
+#### Run the tests
```
Comment (joseluisvaz) [inline README.md:33]: Add instructions to create a venv inside the README.md

File: .idea/other.xml (hunk 1 above)
Comment (joseluisvaz) [inline .idea/other.xml]: Add .idea to .gitignore, and remove this folder form github using git rm

File: cashflows/financial_functions.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import scipy.optimize
+
+
+def xnpv(sorted_flows, rate, year_days=360):
+    npv = 0
+    for flow in sorted_flows:
+        year = (flow.time.value - sorted_flows[0].time.value).days / year_days
+        val = (1 if flow.type == "inflow" else -1) * flow.value * np.power(1 + rate, -year)
+        npv += val
+    return npv
```
Comment (joseluisvaz) [inline cashflows/financial_functions.py]: val = flow.value * np.power(1 + rate, -year)
if flow.type == "outflow":
    val *= -1

File: cashflows/financial_functions.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def xirr(sorted_flows, year_days=360):
+    try:
+        return scipy.optimize.newton(lambda r: xnpv(sorted_flows, r, year_days), 0.0)
```
Comment (joseluisvaz) [inline cashflows/financial_functions.py]: Name of function parameters scipy.optimize.newton(lambda r: ... , x0 = 0.0)

File: cashflows/financial_functions.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    except RuntimeError:
+        return scipy.optimize.brentq(lambda r: xnpv(sorted_flows, r, year_days), -1, 100)
```
Comment (joseluisvaz) [inline cashflows/financial_functions.py]: Same as above

File: cashflows/flows.py [hunk 6]
```diff
@@ -1,20 +1,72 @@
+import warnings
 import numpy as np
 import matplotlib.pyplot as plt
 
+from .financial_functions import *
```
Comment (joseluisvaz) [inline cashflows/flows.py]: Read PEP-8, never do global imports

File: cashflows/flows.py [hunk 7]
```diff
@@ -1,20 +1,72 @@
...
+from dateutil.parser import parse
+from datetime import datetime
```
Comment (joseluisvaz) [inline cashflows/flows.py]: Imports from standard libraries come before imports from user defined libraries.

File: cashflows/financial_functions.py (hunk 3 above)
Comment (joseluisvaz) [reply cashflows/financial_functions.py]: flow.value * (1+rate)**(-year)
if flow.type == "outflow"
    val *= -1

File: cashflows/flows.py [hunk 8]
```diff
@@ -1,10 +1,12 @@
-import warnings
-import numpy as np
 import matplotlib.pyplot as plt
+import numpy as np
+import warnings
```
Comment (joseluisvaz) [inline cashflows/flows.py]: import warnings first, check pep-8 for order of imports

Imports should be grouped in the following order:

standard library imports
related third party imports
local application/library specific imports
You should put a blank line between each group of imports.

File: cashflows/flows.py [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
 
-from .financial_functions import *
+from datetime import datetime, timedelta
 from dateutil.parser import parse
-from datetime import datetime
+
+from .financial_functions import xnpv, xirr
+
```
Comment (joseluisvaz) [inline cashflows/flows.py]: Same as above, send datetime up, and dateutil down

**Review by joseluisvaz — CHANGES_REQUESTED** [review]

File: cashflows/flows.py [hunk 10]
```diff
@@ -1,20 +1,72 @@
...
 
+from .financial_functions import *
+from dateutil.parser import parse
+from datetime import datetime
+
 flow_types = ["inflow", "outflow"]  # TODO: externalize this.
+time_types = ["int", "date"]
+
+
+class Time(object):
```
Comment (joseluisvaz) [inline cashflows/flows.py]: Make an object per class

**Review by joseluisvaz — CHANGES_REQUESTED** [review]

File: cashflows/flows.py (hunk 10 above)
Comment (RHDZMOTA) [reply cashflows/flows.py]: **make a file per class

File: cashflows/flow.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    def __str__(self):
+        return "Flow(type: {this_type}, time: {this_time}, value: {this_value}, currency: {this_currency})".format(
+            this_type=self.type,
+            this_time=self.time.to_string(),
+            this_value=self.value,
+            this_currency=self.currency
+        )
+
+    def to_string(self):
+        return str(self)
```
Comment (joseluisvaz) [inline cashflows/flow.py:23]: erase to string method

File: cashflows/flow.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+
+from .settings import FLOW_TYPES
+from .util import Time
+
+
+class Flow(object):
+    def __init__(self, flow_type, value, time, time_type="int", currency="mxn"):
```
Comment (joseluisvaz) [inline cashflows/flow.py:8]: Add docstring specifing the meaning of the object properties, and the values that they can take

File: cashflows/flow.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        self.type = self.validate_flow_type(flow_type)
+        self.time = Time(time, time_type)
+        self.value = self.validate_value(value)
+        self.currency = self.validate_currency(currency)
+
+    def __str__(self):
+        return "Flow(type: {this_type}, time: {this_time}, value: {this_value}, currency: {this_currency})".format(
+            this_type=self.type,
+            this_time=self.time.to_string(),
```
Comment (joseluisvaz) [inline cashflows/flow.py:17]: self.time.to_string() -> str(self.time)

**Review by joseluisvaz — APPROVED** [review]
Change inflow and outflow to +and -

Change Flow to CashFlow and CashFlow to CashFlowSeries

**Pull request merged**